src/csharp/Grpc.Core/Internal/NativeMetadataCredentialsPlugin.cs

```diff
@@ -78,7 +78,10 @@ private void NativeMetadataInterceptorHandler(IntPtr statePtr, IntPtr serviceUrl
             {
                 var context = new AuthInterceptorContext(Marshal.PtrToStringAnsi(serviceUrlPtr),
                                                          Marshal.PtrToStringAnsi(methodNamePtr));
-                StartGetMetadata(context, callbackPtr, userDataPtr);
+                // Don't await, we are in a native callback and need to return.
+                #pragma warning disable 4014
+                GetMetadataAsync(context, callbackPtr, userDataPtr);
```

Comment:
So what we could do here is to start the operation (and then we need to return control to C core) and handle the continuation via a callback.  But under the hood, async basically creates the same thing (start method and then callback as a continuation), but the compiler rewrites the code so that it handles try-catch  blocks also. So it's easier to let compiler do the hard work for use and use the async method in this way.